src/org/sosy_lab/common/collect/OrderStatisticMapTestSuite.java

```diff
@@ -0,0 +1,369 @@
+        List<Entry<String, String>> pEntries);
+  }
+
+  private OrderStatisticMapFactory factory;
+
+  public OrderStatisticMapTestSuite(OrderStatisticMapFactory pFactory) {
+    factory = pFactory;
+  }
+
+  @SuppressWarnings("unchecked")
+  private OrderStatisticMap<String, String> createMap() {
+    return factory.create(Collections.emptyList());
+  }
+
+  @SuppressWarnings("unchecked")
+  private OrderStatisticMap<String, String> createMap(List<Entry<String, String>> pEntries) {
+    return factory.create(pEntries);
+  }
+
+  private static <K, V> boolean containsEntry(Map<K, V> pMap, Entry<K, V> pEntry) {
+    return pMap.containsKey(pEntry.getKey()) && pEntry.getValue().equals(pMap.get(pEntry.getKey()));
+  }
+
+  private static <K, V> void putEntry(Map<K, V> pMap, Entry<K, V> pEntry) {
+    pMap.put(pEntry.getKey(), pEntry.getValue());
+  }
+
+  private static <K, V> void removeEntry(Map<K, V> pMap, Entry<K, V> pEntry) {
+    Assert.assertEquals(pMap.get(pEntry.getKey()), pMap.remove(pEntry.getKey()));
+  }
+
+  @Test
+  public void testEquals() {
+    OrderStatisticMap<String, String> l1 = createMap();
+    @Var OrderStatisticMap<String, String> l2 = createMap();
+
+    // Check that the map sorts its elements
+    Assert.assertEquals(l1, l2);
+    for (int i = ELEMS.size() - 1; i >= 0; i--) {
+      l2.put(ELEMS.get(i).getKey(), ELEMS.get(i).getValue());
+    }
+    for (int i = 0; i < ELEMS.size(); i++) {
+      l1.put(ELEMS.get(i).getKey(), ELEMS.get(i).getValue());
+    }
+    Assert.assertEquals(l1, l2);
+
+    // Check the map property
+    l2 = createMap();
+    for (int i = ELEMS.size() - 1; i >= 0; i--) {
+      l2.put(ELEMS.get(i).getKey(), ELEMS.get(i).getValue());
+    }
+    for (int i = 0; i < ELEMS.size(); i++) {
+      l2.put(ELEMS.get(i).getKey(), ELEMS.get(i).getValue());
+    }
+    Assert.assertEquals(l1, l2);
+  }
+
+  @Test
+  public void testSerialize() {
+    OrderStatisticMap<String, String> l = createMap();
+    SerializableTester.reserializeAndAssert(l);
+
+    for (int i = 100000; i >= 0; i--) {
+      l.put(String.valueOf(i), String.valueOf(i));
+    }
+    SerializableTester.reserializeAndAssert(l);
+  }
+
+  @Test
+  public void testSubmapView_mutation() {
+    NavigableMap<String, String> map = createMap(ELEMS);
+    NavigableMap<String, String> subMap =
+        map.subMap(ELEMS.get(1).getKey(), true, ELEMS.get(2).getKey(), true);
+
+    Entry<String, String> toAdd = ELEMS_BELOW.get(2);
+
+    putEntry(subMap, toAdd);
+    Assert.assertTrue(containsEntry(subMap, toAdd));
+    Assert.assertTrue(containsEntry(map, toAdd));
+
+    removeEntry(subMap, toAdd);
+    Assert.assertFalse(map.containsKey(toAdd.getKey()));
+    Assert.assertFalse(subMap.containsKey(toAdd.getKey()));
+
+    putEntry(map, toAdd);
+    Assert.assertTrue(containsEntry(subMap, toAdd));
+
+    removeEntry(map, toAdd);
+    Assert.assertFalse(containsEntry(subMap, toAdd));
+  }
+
+  @Test(expected = IllegalArgumentException.class)
```

Comment:
`expected` is bad except in single-statement test cases. Please use `try { ...; fail(); } catch (IllegalArgumentException expected) { }` with as little code inside the `try` as possible.

Reply:
why would you use try/fail instead of an expected excpeption rule? In my opinion this creates much nicer, and especially more readable code. Moreover the nested try-catches should be separate unit-tests, should they?

Reply:
`ExpectedException` has a single advantage over `Test(expected=)`: it lets you reduce the scope where the exception is expected from the whole method to a range of code between the call to ExpectedException and the end of the method. So, it is better than `Test(expected=)`, and if the call to ExpectedException is the second-to-last statement, then it works fine.

However, both methods have one serious problem: If the expected exception is not thrown in the last statement of the method but before, the remainder of the test case will be silently ignored. This is error-prone, and it seems [Google has found out](https://github.com/google/error-prone/issues/219#issuecomment-71244204) that many developers assume that ExpectedException does some magic and actually manages to resume code execution after the expected exception is thrown. So my understanding from following the development of Guava, Truth, and Error-Prone is that both `ExpectedException` and `Test(expected=)` are [not](https://github.com/google/error-prone/issues/219#issuecomment-86110692) [worth](http://errorprone.info/bugpattern/TestExceptionChecker) it in most cases and the old `try-fail-catch` way is [recommended](https://stackoverflow.com/a/38866403/396730) (Christian Gruber is one of the Google developers of these projects) because it is less error-prone *and* allows you to safely write code after the expected exception. Error-Prone also has a check for not forgettting the `fail()` in the `try`, which would be the most important kind of error in this style of writing tests.

All this is true only until we have a newer JUnit that supports writing something like `assertThrows()` with a lambda. I guess then this should be preferred.

Now, independently from this it is of course good if tests are as small as possible, and this could also be improved in this class by splitting them into more methods. And if you follow that strictly, you might decide that the remaining risk of using `ExpectedException` is quite small and that it is fine to use. That would be fine.
For the current set of tests, however, I recommended using `try-fail-catch` because the exception handling is the more important problem than the somewhat large tests (the latter are only a problem for debugging but not necessarily error-prone), and because these tests already use `try-catch` in addition to `Test(expected=)`, switching completely to `try-fail-catch` would be easier.

Reply:
wow, thanks for the detailed response! I never thought about it that way. We never took the actual location where the exception is thrown into account (our unit-tests follow the arrange-act-assert principle and the the to-be-thrown exception is declared in the arrange block). What we do always is check the error message (and by "contract" if any method we implemented throws an exception, it has to have a proper message), and thus we still can be sure that we caught the right exception